### PR TITLE
postgresql@17 17.0 (new formula)

### PR DESCRIPTION
**Pull request by cho-m**
Replace #167634 to work on this for next major release for lower impact.

---

As mentioned in previous PR:

> Another attempt at trying to get extensions working on other PostgreSQL formulae.
> 
> This time, keg directory structure is mostly unchanged other than `datadir` using `#{share}/postgresql` rather than ~~`#{share}/postgresql@16` for PostgreSQL 16~~ `#{share}/postgresql@17` for PostgreSQL 17,

### CODE REVIEW

File: Formula/p/postgresql@17.rb [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    regex(%r{href=["']?v?(17(?:\.\d+)+)/?["' >]}i)
+  end
+
+  keg_only :versioned_formula
+
+  # https://www.postgresql.org/support/versioning/
+  # TODO: deprecate! date: "YYYY-MM-DD", because: :unsupported
+
+  depends_on "docbook" => :build # TODO: Check on this for stable release
+  depends_on "docbook-xsl" => :build # TODO : Check on this for stable release
```
Comment (cho-m) [inline Formula/p/postgresql@17.rb:20]: TODO...

File: Formula/p/postgresql@17.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+  test do
+    system "#{bin}/initdb", testpath/"test" unless ENV["HOMEBREW_GITHUB_ACTIONS"]
+    assert_equal "#{HOMEBREW_PREFIX}/share/#{name}", shell_output("#{bin}/pg_config --sharedir").chomp
+    assert_equal "#{HOMEBREW_PREFIX}/lib/#{name}", shell_output("#{bin}/pg_config --libdir").chomp
+    assert_equal "#{HOMEBREW_PREFIX}/lib/#{name}", shell_output("#{bin}/pg_config --pkglibdir").chomp
+    assert_equal "#{HOMEBREW_PREFIX}/include/#{name}", shell_output("#{bin}/pg_config --pkgincludedir").chomp
+    assert_equal "#{HOMEBREW_PREFIX}/include/#{name}/server",
+                 shell_output("#{bin}/pg_config --includedir-server").chomp
+    assert_match "-I#{Formula["gettext"].opt_include}", shell_output("#{bin}/pg_config --cppflags")
```
Comment (cho-m) [inline Formula/p/postgresql@17.rb:176]: Linux is probably going to fail here as I removed `gettext` dep. Shouldn't be needed based on original reason (libintl.h) as that is in glibc.

File: Formula/p/postgis-pg17.rb [hunk 3]
```diff
@@ -0,0 +1,150 @@
+class PostgisPg17 < Formula
```
Comment (cho-m) [inline Formula/p/postgis-pg17.rb:1]: Wasn't sure on name if we go this route, e.g.
* `postgis-postgresql16` may be like some other formulae (e.g. `qt5`), but a bit lengthy.
* `postgis-pg16` is an abbreviated variant, which is reverse of MacPorts (i.e. `pg16-postgis`)
* `postgis16` could be an option like GTK dependents (where we often add `3` or `4`), but then it looks more like part of actual name (e.g. `libxml2`) which is confusing.

---

NOTE: Probably will end up dropping most of these commits when stable PG 17 is released. Changes are mostly to test that they work for most of our dependents

File: Formula/p/postgresql@17.rb [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+          Find.prune if src.basename.to_s == ".DS_Store"
+          dst.parent.install_symlink src
+        elsif src.directory?
+          dst.mkpath
+        end
+      end
+    end
+
+    # Also link versioned executables
+    bin.each_child { |f| (HOMEBREW_PREFIX/"bin").install_symlink f => "#{f.basename}-#{version.major}" }
```
Comment (cho-m) [inline Formula/p/postgresql@17.rb:134]: Main new logic is here, which will need the most scrutiny.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good so far!

**Review by carlocab — APPROVED** [review]
Seems fine, though the usual keg-only caveat doesn't seem to apply anymore here, since something is linked into HOMEBREW_PREFIX.

**Comment by cho-m** [comment]
> usual keg-only caveat doesn't seem to apply anymore here, since something is linked into HOMEBREW_PREFIX.

Yeah, there isn't really a great way to deal with. Same situation for `rustup`. Currently the only option is extra info in caveats.

---

I will probably merge soon and follow up as necessary. Luckily no one will be upgraded to `postgresql@17` so minimal impact existing users.

Basic usage seemed fine on macOS. Linux is linking via `HOMEBREW_PREFIX/lib/postgresql@17` but didn't hit any CI failure. I can experiment in separate PR if we want an extra RPATH for Cellar or if possible to drop macOS Makefile modification workaround (linking through Cellar is safer, but the special linking dir should always exist after postinstall and will only be an issue when `brew install [--build-bottle | --skip-post-install]`).

**Comment by carlocab** [comment]
> I can experiment in separate PR if we want an extra RPATH for Cellar or if possible to drop macOS Makefile modification workaround (linking through Cellar is safer, but the special linking dir should always exist after postinstall and will only be an issue when `brew install [--build-bottle | --skip-post-install]`).

You could just make their install names `@rpath/` with the appropriate `-Wl,-rpath` flags in `LDFLAGS`. This would skip the prefix relocation and make bottle pouring slightly faster.

**Comment by cho-m** [comment]
> You could just make their install names `@rpath/` with the appropriate `-Wl,-rpath` flags in `LDFLAGS`. This would skip the prefix relocation and make bottle pouring slightly faster.

Can try it out afterward, but a lot the `*FLAGS` are controlled by the Makefile which makes configuring a pain. This was reason for inreplace:
```rb
inreplace "src/Makefile.shlib", "-install_name '$(libdir)/", "-install_name '#{lib}/postgresql/"
```

File: Formula/p/postgresql@17.rb [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  on_linux do
+    depends_on "linux-pam"
+    depends_on "util-linux"
+  end
+
+  def install
+    # Modify Makefile to link macOS binaries using Cellar path. Otherwise, binaries are linked
+    # using #{HOMEBREW_PREFIX}/lib path set during ./configure, which will cause audit failures
+    # for broken linkage as the paths are not created until post-install step.
+    inreplace "src/Makefile.shlib", "-install_name '$(libdir)/", "-install_name '#{lib}/postgresql/"
```
Comment (cho-m) [inline Formula/p/postgresql@17.rb:61]: Looking at RPATHs, following may work:
```suggestion
    inreplace "src/Makefile.shlib", "-install_name '$(libdir)/", "-install_name '@rpath/"
```

With multiple `-rpath` passed in (I think 3 are needed for `@loader_path`, `@loader_path/../lib/postgresql`, and `@loader_path/../../../..`).

Can experiment some more in other PR.

**Pull request merged**